### PR TITLE
endpoint to find with a single call many authorizations for multiple objects and features

### PR DESCRIPTION
**Pull request by corrad82-4s**
## References
* Needed for [angular issue #1044](https://github.com/DSpace/dspace-angular/issues/1044)
* Related to [REST Contract PR](https://github.com/DSpace/RestContract/pull/168)

## Description
We have added an endpoint accepting resource type, and a list of resource uuids and features to which the caller want authorization(s) verified

## Instructions for Reviewers
* Entry point is new method findByObjects in AuthorizationRestRepository class, this class has been refactored a bit. 
* New method collaborates with Utils getBaseObjectRestFromTypeAndUUID method, used to obtain a Rest Object. Starting from list of uuids, Rest Objects are retrieved, and for each objects feature(s) are verified.

List of changes in this PR:
* Refactoring in Utils part retrieving BaseRestObjects in order to have a method retrieving it with apiCategory, model and uuid, informations. This method is used by the existing that retrieves BaseRestObject with uri, and from a new one accepting type ("core.item", "core.collection",...) and uuid and returning its Rest representation.
* A new "findByObject" method has been added to AuthorizationRestRepository
* new private methods in AuthorizationRestRepository to retrieve objects starting from their uuids and evaluating the authorization features for each object
* AuthorizationRestRepository refactoring, added private methods to retrieve authorizations for an object represented by an uri
* Added IT covering same scenarios covered for "/api/authz/authorizations/object" endpoint

**Include guidance for how to test or review your PR.** 
To test it, call the new endpoint "/api/authz/authorizations/objects" with a type parameter ("core.item"), one or many item uuids and one or many feature to be verified. Results should be the same than what obtained by performing single calls for each item uri and for each feature using the single "/api/authz/authorizations/object" endpoint.

## Checklist

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java [hunk 1]
```diff
@@ -154,12 +155,36 @@ public AuthorizationRest findOne(Context context, String id) {
...
+        EPerson user = getUserFromRequestParameter(context, epersonUuid);
+        if (ObjectUtils.notEqual(currUser, user)) {
+            // Temporarily change the Context's current user in order to retrieve
+            // authorizations based on that user
+            context.switchContextUser(user);
+        }
+
+        List<Authorization> authorizations = findAuthorizationsForUri(context, user, uri, featureName);
+
+        if (currUser != user) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java:170]: Is there a reason to use `!=` here when the prior if uses `ObjectUtils.notEqual()`?  It seems like these two ifs should be using the same equality check.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java [hunk 2]
```diff
@@ -168,11 +193,69 @@ public AuthorizationRest findOne(Context context, String id) {
             context.switchContextUser(user);
         }
 
+        List<Authorization> authorizations =
+                findAuthorizationsByUUIDList(context, type, uuidList, user, featureNames);
+
+        if (currUser != user) {
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java:199]: Again, seems like this should use the same equality check as above... `ObjectUtils.notEqual()`, unless I'm misunderstanding something

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java [hunk 3]
```diff
@@ -168,11 +193,69 @@ public AuthorizationRest findOne(Context context, String id) {
...
+        List<BaseObjectRest> objects = uuidList.stream()
+            .map(uuid -> utils.getBaseObjectRestFromTypeAndUUID(context, type, uuid))
+            .collect(Collectors.toList());
+
+        objects.forEach(object ->
+            featureNames.forEach(featureName -> {
+                try {
+                    authorizations.addAll(authorizationsForObject(context, user, featureName, object));
+                } catch (Exception ex) {
+                    log.error(ex.getMessage(), ex);
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java:226]: This should likely provide a more useful error message, or it should just say `log.error(ex);`.  There's no reason to pull out the getMessage() here.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java [hunk 4]
```diff
@@ -964,22 +964,54 @@ public BaseObjectRest getBaseObjectRestFromUri(Context context, String uri) thro
...
+    /**
+     * Gets the rest object of of specified type (i.e. "core.item", "core.collection", "workflow.workflowitem",...)
+     * having given uuid.
+     *
+     * @param context the DSpace context
+     * @param type Object type
+     * @param uuid Object uuid
+     * @return the {@link BaseObjectRest} identified by the provided uuid
+     */
+
```
Comment (tdonohue) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/utils/Utils.java:979]: Tiny thing, there's an unnecessary empty line between the JavaDocs and the method

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 5]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+        getClient(adminToken).perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.community")
+            .param("uuid", comId)
+            .param("uuid", secondComId)
+            .param("projection", "level")
+            .param("embedLevelDepth", "1")
+            .param("feature", alwaysTrue.getName())
+            .param("feature", alwaysFalse.getName())
+            .param("feature", trueForAdmins.getName())
+            .header("X-On-Behalf-Of", eperson.getID()))
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1809]: Just to ensure this call is equivalent to the previous call, we may want to also add the `trueForLoggedInUsers` feature to the list.  That way we prove this works the exact same as the previous one.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 6]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        // verify that it works for administrators - with eperson parameter
+        getClient(adminToken).perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.item")
+            .param("uuid", itemId)
+            .param("uuid", secondItemId)
+            .param("feature", alwaysFalse.getName())
+            .param("feature", alwaysFalse.getName())
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1898]: You've duplicated `alwaysFalse` in several of these calls (this and next two).  It seems like we only need to pass it once?

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 7]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+        // disarm the alwaysThrowExceptionFeature
+        configurationService.setProperty("org.dspace.app.rest.authorization.AlwaysThrowExceptionFeature.turnoff", true);
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        // verify that it works for administrators, no result - with eperson parameter
+        getClient(adminToken).perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.site")
+            .param("uuid", siteId)
+            .param("uuid", siteId)
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1987]: You duplicated this `uuid` param here (and in the next 13 tests). It'd also help make this test clearer if you rename `siteId` to `wrongSiteId` or similar, to make it clear it's invalid

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 8]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+            "boo-invalid",
+            "this-is-not-an-uuid"
+        };
+        // disarm the alwaysThrowExceptionFeature
+        configurationService.setProperty("org.dspace.app.rest.authorization.AlwaysThrowExceptionFeature.turnoff", true);
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+        String epersonToken = getAuthToken(eperson.getEmail(), password);
+//        for (String invalidUri : invalidUris) {
+//            log.debug("findByObjectAndFeatureBadRequestTest - Testing the URI: " + invalidUri);
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:2131]: Please remove commented out debug code

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 9]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+            .param("feature", alwaysTrue.getName()))
+            .andExpect(status().isBadRequest());
+
+        // verify that returns bad request for missing type
+        getClient().perform(get("/api/authz/authorizations/search/objects")
+            .param("uuid", UUID.randomUUID().toString())
+            .param("uuid", UUID.randomUUID().toString())
+            .param("feature", alwaysTrue.getName()))
+            .andExpect(status().isBadRequest());
+//        }
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:2224]: Please remove commented out braces

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 10]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+        String siteId = site.getID().toString();
+
+        // disarm the alwaysThrowExceptionFeature
+        configurationService.setProperty("org.dspace.app.rest.authorization.AlwaysThrowExceptionFeature.turnoff", true);
+
+        // verify that it works for an anonymous user inspecting an admin user - by using the eperson parameter
+        getClient().perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.site")
+            .param("uuid", siteId)
+            .param("uuid", siteId)
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:2245]: Again, `uuid` is unnecessarily duplicated here and in next 3 methods

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 11]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+        context.restoreAuthSystemState();
+        // disarm the alwaysThrowExceptionFeature
+        configurationService.setProperty("org.dspace.app.rest.authorization.AlwaysThrowExceptionFeature.turnoff", true);
+        String anotherToken = getAuthToken(anotherEperson.getEmail(), password);
+
+        // verify that he cannot search the admin authorizations - by using the eperson parameter
+        getClient(anotherToken).perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.site")
+            .param("uuid", siteId)
+            .param("uuid", siteId)
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:2302]: Again `uuid` param duplicated here and in next 3 tests

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 12]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+        Site site = siteService.findSite(context);
+        String siteId = site.getID().toString();
+
+        String adminToken = getAuthToken(admin.getEmail(), password);
+
+        // verify that it works for administrators - with eperson parameter
+        getClient(adminToken).perform(get("/api/authz/authorizations/search/objects")
+            .param("type", "core.site")
+            .param("uuid", siteId)
+            .param("uuid", siteId)
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:2354]: Again `uuid` param unnecessarily duplicated here and in next 4 tests

**Review by tdonohue — CHANGES_REQUESTED** [review]
@corrad82-4s : Gave this a code review today.  Overall, the code looks good and it seems to have valid tests. That said, there's a number of tiny things to cleanup in the code (mostly in the ITs).  Nothing major though.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java (hunk 6 above)
Comment (corrad82-4s) [reply dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1898]: @tdonohue It is to test the case of passing 2 features for which the user is not allowed, and check that the endpoint returns a proper response.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java [hunk 13]
```diff
@@ -154,12 +155,36 @@ public AuthorizationRest findOne(Context context, String id) {
...
+        if (currUser != user) {
+            // restore the real current user
+            context.restoreContextUser();
         }
+        return converter.toRestPage(authorizations, pageable, utils.obtainProjection());
+    }
+
+    @PreAuthorize("#epersonUuid==null || hasPermission(#epersonUuid, 'EPERSON', 'READ')")
+    @SearchRestMethod(name = "objects")
+    public Page<AuthorizationRest> findByObjects(@Parameter(value = "uuid") List<String> uuidList,
```
Comment (Raf-atmire) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java:179]: I'd opt to make the UUID parameter required here. The functionality won't do anything if no uuid is given and it'd just crash due to a nullpointer exception later on anyway I think.
Please make it required, like it was done in the singular findByObject and add a test for this specific case where it's left empty

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java [hunk 14]
```diff
@@ -154,12 +155,36 @@ public AuthorizationRest findOne(Context context, String id) {
...
+            @Parameter(value = "type") String type,
```
Comment (Raf-atmire) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java:180]: Same deal here, I'd also make type required as a request without a type isn't going to be meaningful anyway. However, this will be handled cleanly as you included a check for this later on in the code. This means that implicitely this parameter is required, so we're better off making it explicitely required by setting it here.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 15]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+    /**
+     * Verify that the search by multiple objects and features works properly in allowed scenarios:
+     * - for an administrator
+     * - for an administrator that want to inspect permission of the anonymous users or another user
+     * - for a logged-in "normal" user
+     * - for anonymous
+     *
+     * @throws Exception
+     */
+    public void findByMultipleObjectsAndFeaturesTest() throws Exception {
```
Comment (Raf-atmire) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1538]: In this test, you regularly add/remove features that you're checking. I'd opt to always include all the features that you use throughout the test. This would make it so that the first REST call asks for the same features as the last REST call and that gives us a much cleaner overview of everything that the code handles.

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 16]
```diff
@@ -1517,6 +1525,873 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+    /**
+     * Verify that the find by multiple objects and features
+     * return the 204 No Content code when the requested object doesn't exist but the uri
+     * is potentially valid (i.e. deleted object) or the feature doesn't exist
+     *
+     * @throws Exception
+     */
+    public void findByNotExistingMultipleObjectsAndFeatureTest() throws Exception {
+        UUID wrongSiteId = UUID.randomUUID();
+        String siteId = wrongSiteId.toString();
```
Comment (Raf-atmire) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1972]: Please change this to invalidSiteId or wrongSiteId or something similar

**Review by Raf-atmire — CHANGES_REQUESTED** [review]
Hi @corrad82-4s 
First of all, thank you for the pull request! I've given this a review and added a few comments that I'd like to see changed or hear your feedback about if you don't agree.
Please do let me know if you have any questions or if you don't agree!

Thanks again!

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java (hunk 3 above)
Comment (corrad82-4s) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/AuthorizationRestRepository.java]: hi @tdonohue ok, I have added an error message stating the error while retrieving the authorizations. I think having the exception message as part of this message, and then the whole exception thrown in following can better indicate what happened.

**Comment by corrad82-4s** [comment]
Hi @tdonohue , @benbosman thank you for your review and suggestions, I have pushed a couple of commit with suggested changes, apart than what reported in this comment https://github.com/DSpace/DSpace/pull/7955#discussion_r719113025

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java [hunk 17]
```diff
@@ -1517,6 +1527,876 @@ public void findByObjectAndFeatureInternalServerErrorTest() throws Exception {
...
+                    )),
+                    hasJsonPath("$.type", is("authorization")),
+                    hasJsonPath("$._embedded.feature.id", is(trueForLoggedUsers.getName())),
+                    hasJsonPath("$._embedded.eperson.id", is(eperson.getID().toString()))
+                )
+            )));
+
+        // verify that it works for administators inspecting other users - by assuming login
+        getClient(adminToken).perform(baseFeatureRequest.get()
+//            .param("feature", trueForLoggedUsers.getName())
```
Comment (tdonohue) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/AuthorizationRestRepositoryIT.java:1829]: Tiny thing, you can remove this.  It's already in `baseFeatureRequest`

**Review by tdonohue — APPROVED** [review]
👍 Thanks for the updates, @corrad82-4s !  The code and tests look good to me now.  I built and ran a few brief tests and it appears to work (I also reviewed the ITs to verify they test it better than I could manually).   One very minor thing inline below...but this looks good to merge!

**Review by Raf-atmire — APPROVED** [review]
Hi @corrad82-4s Thank you for addressing the feedback! This looks good and ready to be merged, approval from me. Thank you for the work!
